### DISCUSSION TITLE
PreloadWhere Documentation/Examples

### DISCUSSION THREAD

**Discussion by daddz** [stephenafamo/bob#172] [Q&A] [answered]
I am trying to figure out how I can use the generated code to filter a select based on a value in a related to-one table.

While digging through the repository I found the `PreloadWhere` option which looks like it could be used but unfortunately it's not really documented and there are no examples on how to use it.

Could you show/add some documentation/examples on how to use it with generated `SelectWhere`s?

**Comment by stephenafamo** [comment, marked as answer]
The best way to do this at the moment is to do a fresh join and then filter as you would:

```go
	videos, err := models.Videos.Query(ctx, db,
		models.PreloadVideoUser(),
		models.SelectJoins(ctx).Videos.InnerJoin.User,
		models.SelectWhere.Users.ID.GT(200),
	).One()
```

The generated query will has 2 joins, but the DB is smart enough to do the work just once. You can verify this using `EXPLAIN ANALYZE`

____________________

### Why can't `PreloadX` take `WhereMods`?

The reason is that to preload, a random suffix is added to the name of the table. This isolates it from the rest of the query and makes it extremely unlikely that it conflicts with any other part of the query.

So, if you wanted to filter using the same join that was generated for preloading, then we would need to pass the generated table name back to the where mod.  
I could not find a good way to do this that did not ruin the user experience.

**Reply by daddz** [reply]
Ah thanks, I also got a working variant with the extra join.

Would it make sense to expose the generated alias from the `Preloader` maybe?

Something along those lines (ugly but should get the idea across):
```go
preloader := models.PreloadFooBar()
alias := preloader.Alias()

filter := sm.Where(psql.Quote(fmt.Sprintf("%s.%s", alias, models.ColumnNames.Bar.FieldIWantToFilterOn)).EQ(psql.Arg("myval")))
```

**Reply by stephenafamo** [reply]
I did consider this, but as you can see, it isn't the best user experience. I would prefer something more elegant.

**Reply by daddz** [reply]
Thanks for your help, the additional join will do for now!

**Reply by daddz** [reply]
Hey @stephenafamo, I ran into another problem with your suggested solution: It doesn't work if there are multiple relations to the same table. 

Example:
```go
// created_by and edited_by are both fks to the same table
models.Videos.Query(ctx, db,
      models.SelectJoins(ctx).Videos.InnerJoin.UploadedBy,
      models.SelectJoins(ctx).Videos.InnerJoin.EditedBy).All()
```

To workaround that issue I tried to add a new PreloadOption that let's me override the alias: https://github.com/daddz/bob/commit/f546eabba5fc329d9bca168cc828b556c9f80682

I think the user experience is not as bad as in the previous idea:
```go
models.Videos.Query(ctx, db,
  models.PreloadVideoUploadedByUser(psql.PreloadAs("uploaded_by"),
  models.PreloadVideoEditedByUser(psql.PreloadAs("edited_by"),
  sm.Where(psql.Quote("uploaded_by", models.ColumnNames.User.Name).EQ(psql.Arg("john")))
  sm.OrderBy(psql.Quote("edited_by", models.ColumnNames.User.ID).Desc())
```

What do you think about this approach?

This could be further improved if there would be a way to alias the auto-generated `SelectWhere`s:

```go
models.Videos.Query(ctx, db,
  models.PreloadVideoUploadedByUser(psql.PreloadAs("uploaded_by"),
  models.PreloadVideoEditedByUser(psql.PreloadAs("edited_by"),
  models.SelectWhere.Users.As("uploaded_by").Name.EQ("john"),
  sm.OrderBy(models.ColumnNames.Users.As("edited_by").ID).Desc())
```

**Reply by stephenafamo** [reply]
This is a really nice idea. 

Can you send in a PR (possibly with additional documentation)

**Reply by daddz** [reply]
Sure, I'll send one later!

Do you have an idea on how the auto-generated SelectWhere/ColumnNames could be made aliasable easily?

**Reply by stephenafamo** [reply]
The alias for Postgres is chosen here https://github.com/stephenafamo/bob/blob/v0.26.1/dialect/psql/load.go#L66 (mysql and sqlite are similar)

So if you add the option, you can then check for it there

**Reply by daddz** [reply]
Ah no that's not what I meant, I have that already: https://github.com/daddz/bob/commit/f546eabba5fc329d9bca168cc828b556c9f80682#diff-e8a6434756d2e9bc524e5ab19bb5a585f162c46d953e468186e9fc851e96f08dR71

I was referring to the second part of the idea regarding making it convenient to use with the autogenerated code - namely SelectWhere and ColumnNames.

Do you have an idea on how a design could look like and where it would make sense to add (e.g. in the template, expression or mod)?
Currently it would look like this:
```go
...
sm.Where(psql.Quote("uploaded_by", models.ColumnNames.User.Name).EQ(psql.Arg("john"))),
sm.OrderBy(psql.Quote("edited_by", models.ColumnNames.User.ID).Desc()).All()
```

Whereas something like this would be more convenient:
```go
...
models.SelectWhere.Users.As("uploaded_by").Name.EQ("john"),
sm.OrderBy(models.ColumnNames.Users.As("edited_by").ID).Desc())
```

Or maybe something in this direction:
```go
uploadedBy := models.PreloadVideoUploadedByUser(psql.PreloadAs("uploaded_by"))
editedBy := models.PreloadVideoCreatedByUser(psql.PreloadAs("edited_by"))

models.Videos.Query(ctx, db,
  uploadedBy,
  editedBy,
  models.SelectWhere.Users.As(uploadedBy).Name.EQ("john),
  sm.OrderBy(models.ColumnNames.Users.As(editedBy).ID).Desc()).All()
```

In this case I was thinking about an `.As()` method on the `models.SelectWhere.Users` and `models.ColumnNames.Users` but not sure if that would make sense or there is a prettier way.

**Reply by stephenafamo** [reply]
I can take a look at that part.

**Reply by stephenafamo** [reply]
I created a PR in #221 

I decided to name the method `AliasedAs()`. While this is more verbose, it properly communicates what is happening and has a much less chance of conflicting with a column name (which will lead to compile errors).

**Reply by stephenafamo** [reply]
I've also added the `AliasedAs()` method to the generated Join mods in #222

**Reply by daddz** [reply]
So far I tried the changes from #221 and it's really nice! Makes the usability a lot smoother, thank you!

**Comment by daddz** [comment]
Ah sorry, should have opened a discussion instead of an issue.

I played around with `PreloadWhere` a bit more and it doesn't seem to be what I would need in this case. It adds the expressions to the join condition.

I tried to put what I want to achieve in an example query:
```
SELECT 
  "foo"."id" AS "id",
  "bar_1234"."id" AS "bar_1234.id"
  "bar_1234"."foo_id" AS "bar_1234.foo_id"
FROM "foo" AS "foo"
  LEFT JOIN "bar" as "bar_1234" 
    ON ("foo"."id" = "bar_1234"."foo_id") --- up until here is what comes out of models.PreloadFooBar()
WHERE
  "bar_1234"."field_i_want_to_filter_on" = 1 --- this is what I'd like to add
```

**Reply by stephenafamo** [reply]
It should work like this:

```go
	foos, err := models.Foos.Query(ctx, db,
		models.PreloadFooBar(),
		models.SelectJoins(ctx).Foos.LeftJoin.Bar,
		models.SelectWhere.Bar.Field.EQ("value"),
	).One()
```

This should both preload the `bar` for every returned `foo` **AND** return only `foos` that match the condition.
